File: packages/scripts/generate/generate-invalid-decision-embeddings.ts
/**
 * 为专利无效决定生成向量嵌入
 *
 * 功能：
 * 1. 创建 patent_decisions_v2_embeddings 表
 * 2. 分块处理无效决定文档
 * 3. 使用 BGE-M3 生成向量嵌入
 * 4. 批量插入数据库
 */

import { Pool } from 'pg'
import { EmbeddingAdapter } from '@yunpat/core'

// 配置
const CONFIG = {
  pgHost: 'localhost',
  pgPort: 5432,
  pgDatabase: 'legal_world_model',
  pgUser: 'postgres',
  pgPassword: 'nxLVXyZ3e87L0kE8Xqx3AB9NK1z74pwOdjugqpc7hc',

  // 分块配置
  chunkSize: 500, // 每块字符数
  chunkOverlap: 50, // 重叠字符数
  maxChunks: 10, // 每个文档最大分块数

  // 批处理配置
  batchSize: 100, // 每批处理文档数
  delayMs: 1000, // 批次间延迟（毫秒）
}

/**
 * 创建向量表
 */
async function createEmbeddingsTable(pool: Pool) {
  console.log('[Step 1] 创建向量表...')

  await pool.query(`
    CREATE TABLE IF NOT EXISTS patent_decisions_v2_embeddings (
      id SERIAL PRIMARY KEY,
      document_id UUID NOT NULL,
      document_number VARCHAR(255),
      chunk_index INTEGER,
      chunk_text TEXT NOT NULL,
      vector VECTOR(1024) NOT NULL,
      weight DOUBLE PRECISION DEFAULT 1.0,
      created_at TIMESTAMP DEFAULT NOW(),

      CONSTRAINT fk_document
        FOREIGN KEY (document_id)
        REFERENCES patent_decisions_v2(id)
        ON DELETE CASCADE
    );
  `)

  // 创建索引（稍后在数据插入后创建）
  console.log('✅ 向量表创建成功')
}

/**
 * 分块处理文本
 */
function chunkText(text: string, chunkSize: number, overlap: number): string[] {
  const chunks: string[] = []

  // 按段落分割
  const paragraphs = text.split(/\n\n+/)
  let currentChunk = ''

  for (const paragraph of paragraphs) {
    if (currentChunk.length + paragraph.length > chunkSize) {
      if (currentChunk.length > 0) {
        chunks.push(currentChunk.trim())
        currentChunk = paragraph.slice(-overlap) // 保留重叠部分
      } else {
        // 单个段落太长，强制分割
        for (let i = 0; i < paragraph.length; i += chunkSize - overlap) {
          chunks.push(paragraph.slice(i, i + chunkSize))
        }
      }
    } else {
      currentChunk += (currentChunk ? '\n\n' : '') + paragraph
    }
  }

  if (currentChunk.length > 0) {
    chunks.push(currentChunk.trim())
  }

  return chunks.slice(0, CONFIG.maxChunks)
}

/**
 * 生成向量嵌入
 */
async function generateEmbeddings(pool: Pool) {
  console.log('[Step 2] 生成向量嵌入...')

  // 1. 初始化 BGE-M3 嵌入模型
  console.log('初始化 BGE-M3 嵌入模型...')
  const embedding = await EmbeddingAdapter.createBGEEmbedding({
    model: 'BAAI/bge-m3',
    encodingFormat: 'float',
  })

  // 2. 查询所有无效决定
  const { rows: documents } = await pool.query(
    'SELECT id, document_number, content FROM patent_decisions_v2 ORDER BY id'
  )

  console.log(`找到 ${documents.length} 个无效决定文档`)

  // 3. 批量处理
  let processed = 0
  let totalChunks = 0

  for (let i = 0; i < documents.length; i += CONFIG.batchSize) {
    const batch = documents.slice(i, i + CONFIG.batchSize)
    console.log(
      `\n处理批次 ${Math.floor(i / CONFIG.batchSize) + 1}/${Math.ceil(documents.length / CONFIG.batchSize)}`
    )

    // 准备批次数据
    const embeddingsData: Array<{
      documentId: string
      documentNumber: string
      chunkIndex: number
      chunkText: string
      vector: number[]
    }> = []

    for (const doc of batch) {
      // 分块
      const chunks = chunkText(doc.content, CONFIG.chunkSize, CONFIG.chunkOverlap)
      console.log(`  文档 ${doc.document_number}: ${chunks.length} 个分块`)

      // 生成向量嵌入
      for (let j = 0; j < chunks.length; j++) {
        try {
          const embedding = await embedding.embedText(chunks[j])

          embeddingsData.push({
            documentId: doc.id,
            documentNumber: doc.document_number,
            chunkIndex: j,
            chunkText: chunks[j],
            vector: embedding,
          })
        } catch (err) {
          console.error(`    ❌ 分块 ${j} 嵌入生成失败:`, err)
        }
      }

      processed++
    }

    // 批量插入数据库
    console.log(`  插入 ${embeddingsData.length} 个向量...`)
    await insertEmbeddingsBatch(pool, embeddingsData)

    totalChunks += embeddingsData.length
    console.log(`✅ 进度: ${processed}/${documents.length} 文档, ${totalChunks} 个向量`)

    // 延迟，避免 API 限流
    if (i + CONFIG.batchSize < documents.length) {
      console.log(`  等待 ${CONFIG.delayMs}ms...`)
      await sleep(CONFIG.delayMs)
    }
  }

  console.log(`\n✅ 向量嵌入生成完成: ${totalChunks} 个向量`)
}

/**
 * 批量插入向量嵌入
 */
async function insertEmbeddingsBatch(
  pool: Pool,
  data: Array<{
    documentId: string
    documentNumber: string
    chunkIndex: number
    chunkText: string
    vector: number[]
  }>
) {
  const client = await pool.connect()

  try {
    await client.query('BEGIN')

    for (const item of data) {
      await client.query(
        `
        INSERT INTO patent_decisions_v2_embeddings
          (document_id, document_number, chunk_index, chunk_text, vector)
        VALUES ($1, $2, $3, $4, $5::vector(1024))
        `,
        [
          item.documentId,
          item.documentNumber,
          item.chunkIndex,
          item.chunkText,
          `[${item.vector.join(',')}]`,
        ]
      )
    }

    await client.query('COMMIT')
  } catch (err) {
    await client.query('ROLLBACK')
    throw err
  } finally {
    client.release()
  }
}

/**
 * 创建向量索引
 */
async function createVectorIndex(pool: Pool) {
  console.log('\n[Step 3] 创建向量索引...')

  // 计算索引参数
  const { rows } = await pool.query('SELECT COUNT(*) FROM patent_decisions_v2_embeddings')
  const totalRows = parseInt(rows[0].count)
  const lists = Math.floor(Math.sqrt(totalRows))

  console.log(`总向量数: ${totalRows}, 索引参数: lists = ${lists}`)

  // 创建 IVFFlat 索引
  await pool.query(
    `
    CREATE INDEX IF NOT EXISTS idx_patent_decisions_v2_embeddings_vector
    ON patent_decisions_v2_embeddings
    USING ivfflat (vector vector_cosine_ops)
    WITH (lists = $1);
    `,
    [lists]
  )

  console.log('✅ 向量索引创建成功')
}

/**
 * 主函数
 */
async function main() {
  console.log('========================================')
  console.log('专利无效决定向量嵌入生成器')
  console.log('========================================\n')

  const pool = new Pool({
    host: CONFIG.pgHost,
    port: CONFIG.pgPort,
    database: CONFIG.pgDatabase,
    user: CONFIG.pgUser,
    password: CONFIG.pgPassword,
    max: 10,
  })

  try {
    // Step 1: 创建表
    await createEmbeddingsTable(pool)

    // Step 2: 生成向量嵌入
    await generateEmbeddings(pool)

    // Step 3: 创建索引
    await createVectorIndex(pool)

    console.log('\n========================================')
    console.log('✅ 全部完成！')
    console.log('========================================')

    // 显示统计
    const { rows } = await pool.query('SELECT COUNT(*) FROM patent_decisions_v2_embeddings')
    console.log(`总向量数: ${rows[0].count}`)
  } catch (err) {
    console.error('\n❌ 错误:', err)
    process.exit(1)
  } finally {
    await pool.end()
  }
}

/**
 * 延迟函数
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

// 运行
main()

File: packages/scripts/generate/generate_invalid_decision_embeddings.py
#!/usr/bin/env python3
"""
为专利无效决定生成向量嵌入（Python 版本）

使用 BGE-M3 模型生成向量嵌入，并存储到 PostgreSQL
"""

import psycopg2
import numpy as np
from sentence_transformers import SentenceTransformer
from typing import List, Tuple
import time

# 配置
CONFIG = {
    'pg_host': 'localhost',
    'pg_port': 5432,
    'pg_database': 'legal_world_model',
    'pg_user': 'postgres',
    'pg_password': 'nxLVXyZ3e87L0kE8Xqx3AB9NK1z74pwOdjugqpc7hc',

    # 测试模式
    'test_mode': True,
    'max_documents': 10,

    # 分块配置
    'chunk_size': 500,
    'chunk_overlap': 50,
    'max_chunks': 10,
}

def chunk_text(text: str, chunk_size: int, overlap: int, max_chunks: int) -> List[str]:
    """分块处理文本"""
    chunks = []
    paragraphs = text.split('\n\n')
    current_chunk = ''

    for paragraph in paragraphs:
        if len(current_chunk) + len(paragraph) > chunk_size:
            if current_chunk:
                chunks.append(current_chunk.strip())
                current_chunk = paragraph[-overlap:] if len(paragraph) > overlap else paragraph
            else:
                # 单个段落太长，强制分割
                for i in range(0, len(paragraph), chunk_size - overlap):
                    chunks.append(paragraph[i:i + chunk_size])
        else:
            current_chunk += ('\n\n' if current_chunk else '') + paragraph

    if current_chunk:
        chunks.append(current_chunk.strip())

    return chunks[:max_chunks]

def main():
    print("=" * 40)
    print("专利无效决定向量嵌入生成器（Python 版本）")
    print("=" * 40)
    print(f"\n测试模式: 只处理前 {CONFIG['max_documents']} 个文档\n")

    # 连接数据库
    print("[Step 1] 连接数据库...")
    conn = psycopg2.connect(
        host=CONFIG['pg_host'],
        port=CONFIG['pg_port'],
        database=CONFIG['pg_database'],
        user=CONFIG['pg_user'],
        password=CONFIG['pg_password']
    )
    cursor = conn.cursor()
    print("✅ 数据库连接成功")

    # 创建表
    print("\n[Step 2] 创建向量表...")
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS patent_decisions_v2_embeddings (
            id SERIAL PRIMARY KEY,
            document_id UUID NOT NULL,
            document_number VARCHAR(255),
            chunk_index INTEGER,
            chunk_text TEXT NOT NULL,
            vector VECTOR(1024) NOT NULL,
            weight DOUBLE PRECISION DEFAULT 1.0,
            created_at TIMESTAMP DEFAULT NOW()
        );
    """)
    conn.commit()
    print("✅ 向量表创建成功")

    # 加载 BGE-M3 模型
    print("\n[Step 3] 加载 BGE-M3 模型...")
    model = SentenceTransformer('BAAI/bge-m3')
    print("✅ BGE-M3 模型加载成功")

    # 查询文档
    print("\n[Step 4] 查询文档...")
    limit = CONFIG['max_documents'] if CONFIG['test_mode'] else 9562
    cursor.execute(
        "SELECT id, document_number, content FROM patent_decisions_v2 ORDER BY id LIMIT %s",
        (limit,)
    )
    documents = cursor.fetchall()
    print(f"✅ 找到 {len(documents)} 个文档")

    # 生成向量嵌入
    print("\n[Step 5] 生成向量嵌入...")
    total_chunks = 0
    start_time = time.time()

    for i, (doc_id, doc_number, content) in enumerate(documents, 1):
        print(f"\n处理文档 {i}/{len(documents)}: {doc_number}")

        # 分块
        chunks = chunk_text(content, CONFIG['chunk_size'], CONFIG['chunk_overlap'], CONFIG['max_chunks'])
        print(f"  分块数: {len(chunks)}")

        # 生成向量嵌入
        for j, chunk in enumerate(chunks):
            try:
                print(f"    生成向量 {j + 1}/{len(chunks)}...", end='', flush=True)

                # 生成向量
                embedding = model.encode(chunk, normalize_embeddings=True)

                # 插入数据库
                cursor.execute(
                    """
                    INSERT INTO patent_decisions_v2_embeddings
                      (document_id, document_number, chunk_index, chunk_text, vector)
                    VALUES (%s, %s, %s, %s, %s)
                    """,
                    (doc_id, doc_number, j, chunk, embedding.tolist())
                )

                total_chunks += 1
                print(" ✅")

            except Exception as err:
                print(f" ❌ 失败: {err}")

        # 每 5 个文档提交一次
        if i % 5 == 0:
            conn.commit()

    # 最终提交
    conn.commit()

    # 统计
    elapsed = time.time() - start_time
    print(f"\n✅ 向量嵌入生成完成: {total_chunks} 个向量")
    print(f"   总耗时: {elapsed:.1f} 秒")
    print(f"   平均速度: {total_chunks / elapsed:.2f} 向量/秒")

    # 查询统计
    cursor.execute("SELECT COUNT(*) FROM patent_decisions_v2_embeddings")
    count = cursor.fetchone()[0]
    print(f"\n总向量数: {count}")

    print("\n" + "=" * 40)
    print("✅ 测试完成！")
    print("=" * 40)

    # 关闭连接
    cursor.close()
    conn.close()

if __name__ == '__main__':
    main()
